tests/cli/shm_ui.c

```diff
@@ -0,0 +1,471 @@
+        // Advance line pointer for the next joystick
+        wmove(GAMEPAD_WIN, ++line, 0);
+    }
+
+    // Move to buttons section
+    line += 2;
+    wmove(GAMEPAD_WIN, line, 0);
+
+    // Print pressed buttons
+    mvwprintw(GAMEPAD_WIN, line++, 1, "Pressed Buttons:");
+    wmove(GAMEPAD_WIN, line, 0);
+    for (int i = 0; i < NUM_GAMEPAD_BUTTONS; i++) {
+        // Move to button row and clear previous entry
+        wclrtoeol(GAMEPAD_WIN);
+        // Show button name and an "X" next to it if pressed
+        mvwprintw(GAMEPAD_WIN, line, 5, "%s", button_names[i]);
+        if (pressed_buttons & (1 << i)) {
+            mvwprintw(GAMEPAD_WIN, line, BUTTON_PRESSED_FLAG_COL, "X");
+        }
+        // Move to next button
+        wmove(GAMEPAD_WIN, ++line, 0);
+    }
+
+    // Draw box and refresh
+    box(GAMEPAD_WIN, 0, 0);
+    wrefresh(GAMEPAD_WIN);
+}
+
+/**
+ * Displays a device's current state of parameters in a formatted table.
+ * NET (NET_HANDLER) and EXE (EXECUTOR) refers to whether the parameter is subscribed by the process.
+ * Param Idx (int) | Name (str) | NET (bool) | EXE (bool) | Command (var) | Data (var)
+ * 
+ * Arguments:
+ *    catalog: current shared memory catalog; Used to handle when device at shm_idx is invalid
+ *    dev_ids: current shared memory device information; used to get device information at shm_idx
+ *    shm_idx: the index of shared memory of the device to display
+ *             set to MAX_DEVICES if displaying the custom data block is desired
+ */
+void display_device(uint32_t catalog, dev_id_t dev_ids[MAX_DEVICES], int shm_idx) {
+    // Special case handling
+    const int show_custom_data = (shm_idx == MAX_DEVICES);
+    if (!show_custom_data && !(catalog & (1 << shm_idx))) { // Device is not connected at this shared memory index
+        // Clear the window if not clear already (Happens when we disconnect a device while we're inspecting it)
+        if (!DEVICE_WIN_IS_BLANK) {
+            // Clear the entire window, but put back the header and the borders
+            wclear(DEVICE_WIN);
+            mvwprintw(DEVICE_WIN, 1, 1, DEVICE_WIN_HEADER);
+            mvwprintw(DEVICE_WIN, DEVICE_HEIGHT - 2, 1, NOTE);
+            box(DEVICE_WIN, 0, 0);
+            wrefresh(DEVICE_WIN);
+            DEVICE_WIN_IS_BLANK = 1; // Flag that the device window is cleared
+        }
+        // Return because there is no data to display for a disconnected device
+        return;
+    } else if (DEVICE_WIN_IS_BLANK) { // Window was previously clear (i.e Switching from no device to a connected device)
+        // Put back title and box
+        mvwprintw(DEVICE_WIN, 1, 1, DEVICE_WIN_HEADER);
+        mvwprintw(DEVICE_WIN, DEVICE_HEIGHT - 2, 1, NOTE);
+        box(DEVICE_WIN, 0, 0);
+        wrefresh(DEVICE_WIN);
+        // Note that we will display data, so DEVICE_WIN will no longer be "empty"
+        DEVICE_WIN_IS_BLANK = 0;
+    } else { 
+        // Switching from one connected device to another
```

Comment:
Why have this empty `else` block?

Reply:
It's there only for clarity to understand the logic if someone else were to make changes to the UI in the future.